In [1]:
%matplotlib inline

import pandas as pd
import matplotlib.pyplot as plt
import matplotlib
import numpy as np
from io import StringIO

matplotlib.style.use('ggplot') 
#pd.set_option('display.max_columns', None)


# County, State, National Data

In [2]:
state_abbrev = {
    'Alabama': 'AL','Alaska': 'AK','Arizona': 'AZ','Arkansas': 'AR','California': 'CA',
    'Colorado': 'CO','Connecticut': 'CT','Delaware': 'DE','Florida': 'FL','Georgia': 'GA',
    'Hawaii': 'HI','Idaho': 'ID','Illinois': 'IL','Indiana': 'IN','Iowa': 'IA','Kansas': 'KS',
    'Kentucky': 'KY','Louisiana': 'LA','Maine': 'ME','Maryland': 'MD','Massachusetts': 'MA',
    'Michigan': 'MI','Minnesota': 'MN','Mississippi': 'MS','Missouri': 'MO','Montana': 'MT',
    'Nebraska': 'NE','Nevada': 'NV','New Hampshire': 'NH','New Jersey': 'NJ','New Mexico': 'NM',
    'New York': 'NY','North Carolina': 'NC','North Dakota': 'ND','Ohio': 'OH','Oklahoma': 'OK',
    'Oregon': 'OR','Pennsylvania': 'PA','Rhode Island': 'RI','South Carolina': 'SC',
    'South Dakota': 'SD','Tennessee': 'TN','Texas': 'TX','Utah': 'UT','Vermont': 'VT',
    'Virginia': 'VA','Washington': 'WA','West Virginia': 'WV','Wisconsin': 'WI','Wyoming': 'WY',
    'District of Columbia': 'DC'
}

In [4]:
# Comparison Data from David Leip's Elections Atlas
# Source: http://uselectionatlas.org/RESULTS/

# 2004
# George W. Bush	Richard Cheney	Republican	62,039,572	50.73%	286	53.2%  
# John Kerry	John Edwards	Democratic	59,027,115	48.26%	251	46.7%  
# Total	122,303,590		538	

# 2008
# Barack H. Obama	Joseph R. Biden, Jr.	Democratic	69,499,428	52.86%	365	67.8%  
# John S. McCain, III	Sarah H. Palin	Republican	59,950,323	45.60%	173	32.2%  
# Total	131,473,705		538	

# 2012
# Barack H. Obama	Joseph R. Biden, Jr.	Democratic	65,918,507	51.01%	332	61.7%  
# Willard Mitt Romney	Paul Ryan	Republican	60,934,407	47.15%	206	38.3%  
# Total	129,237,642		538	

# 2016
# Donald J. Trump	Michael R. Pence	Republican	62,985,106	45.94%	304	56.5%  
# Hillary Clinton	Tim Kaine	Democratic	65,853,625	48.03%	227	42.2%  
# Total	137,100,229		538	


In [3]:
# Load the 2012, 2016 Voting and Turnout data
# Source: https://github.com/kyaroch/2012_and_2016_presidential_election_results_by_county

voting1_df = pd.read_csv('./data/2012_and_2016_presidential_election_results_by_county.csv',
                         converters={'fips_code': lambda x: str(x)})

columns = ['state_postal_abbrev', 'fips_code', 'county_name', '2012_total_votes', 
           'obama_votes', 'romney_votes', '2016_total_votes', 'clinton_votes', 
           'trump_votes', '2012_adult_pop', '2016_extrapolated_adult_pop', 
           '2012_turnout', '2016_turnout']

renames = ['state',  'fips_code', 'county_name',  'total_2012', 'dem_2012', 
           'rep_2012','total_2016', 'dem_2016', 'rep_2016', 'vap_2012','vap_2016',
           'turnout_2012','turnout_2016']

voting1_df = voting1_df[columns]
voting1_df.rename(columns = dict(zip(columns, renames)), inplace=True)

#Fix fips_code, add leading zero
voting1_df['fips_code'] = voting1_df['fips_code'].apply(lambda x: '0' + x if len(x) < 5 else x)

voting1_df['county_name'] = voting1_df['county_name'] + " County"

# voting1_df[voting1_df['2016_results_official'] == 0].count()
# 385 unofficial results

#Close enough to above numbers from Leip's election atlas. 
print(voting1_df.sum(numeric_only=True).apply(lambda x: '%.2f' % x))

#Turn year into column
year_dfs = []

for year in [2012, 2016]:
    cols = 'state fips_code county_name total_{0} dem_{0} rep_{0} vap_{0} turnout_{0}'.format(year)
    cols = cols.split(' ')
    temp_df = voting1_df[cols].copy()
    temp_df['year'] = year
    renames = dict(zip(cols, ['state', 'fips_code', 'county_name',
                              'county_num', 'dem_num', 'rep_num', 'vap', 'turnout']))
    temp_df.rename(columns=renames, inplace=True)
    year_dfs.append(temp_df)

voting1_df = pd.concat(year_dfs)

voting1_df['dem_margin'] = (voting1_df['dem_num'] - voting1_df['rep_num'])/voting1_df['county_num']

#voting1_df.sort_values(by='dem_margin')
voting1_df.head()


total_2012      128957555.00
dem_2012         65791440.00
rep_2012         60767069.00
total_2016      136352670.00
dem_2016         65689549.00
rep_2016         62778232.00
vap_2012        239849637.00
vap_2016        249735736.00
turnout_2012         1751.15
turnout_2016         1775.76
dtype: object


state fips_code     county_name  county_num  dem_num  rep_num     vap  \
0    AL     01001  Autauga County       23973     6363    17379   40874   
1    AL     01003  Baldwin County       85491    18424    66016  147416   
2    AL     01005  Barbour County       11517     5912     5550   21334   
3    AL     01007     Bibb County        8420     2202     6132   17796   
4    AL     01009   Blount County       24060     2970    20757   43876   

   turnout  year  dem_margin  
0   0.5865  2012   -0.459517  
1   0.5799  2012   -0.556690  
2   0.5398  2012    0.031432  
3   0.4731  2012   -0.466746  
4   0.5484  2012   -0.739277

In [4]:
# Load Voting Age Population 2005 to 2009 average, can't find yearly VAP data
# Source: https://www.census.gov/rdo/data/voting_age_population_by_citizenship_and_race_cvap.html

# 'latin-1' deals with e-acute: https://stackoverflow.com/questions/5552555
population2009_df = pd.read_csv('./data/CVAP_CSV_Format_2005-2009/County.csv', encoding='latin-1')

population2009_df['fips_code'] = population2009_df['GEOID'].apply(lambda x: x[7:])
population2009_df = population2009_df[(population2009_df['LNTITLE'] == 'Total')]
population2009_df.rename(columns={'CVAP_EST':'vap'}, inplace=True)
population2009_df = population2009_df[['fips_code', 'vap']]  

population2009_df.head()

fips_code     vap
0      01001   35315
13     01003  128945
26     01005   22285
39     01007   16255
52     01009   41095

In [5]:
# Load 2004-2012 Voting data 
# Source: https://github.com/helloworlddata/us-presidential-election-county-results
columns = ['year', 'state', 'county', 'fips', 'vote_rep', 'vote_dem', 'vote_total']
renames = ['year', 'state', 'county_name', 'fips_code', 'rep_num', 'dem_num', 'county_num']


voting2_df = pd.read_csv('./data/us-presidential-election-county-results-2004-through-2012.csv')

voting2_df = voting2_df[columns]
voting2_df.rename(columns=dict(zip(columns, renames)), inplace=True)

# Fix Laclede County Results, otherwise state switches to D
# https://en.wikipedia.org/wiki/United_States_presidential_election_in_Missouri,_2008
# Laclede	32.4%	5,218	67.6%	10,875	16,093
voting2_df.loc[(voting2_df['state'] == 'MO') & 
           (voting2_df['year'] == 2008) & 
           (voting2_df['county_name'] == 'Laclede County'), 'county_num'] = 16093

voting2_df['dem_margin'] = (voting2_df['dem_num'] - voting2_df['rep_num'])/voting2_df['county_num'] 

voting2_df = voting2_df[(voting2_df['year'].isin([2004, 2008])) &
                        (voting2_df['state'] != 'AK')]

voting2_df = pd.merge(voting2_df, population2009_df, on='fips_code')

voting2_df['turnout'] = voting2_df['county_num'] / voting2_df['vap']

#Close enough to above numbers from Leip's election atlas. 
#print(voting2_df[.sum(numeric_only=True).apply(lambda x: '%.2f' % x))


voting2_df.head()

year state     county_name fips_code  rep_num  dem_num  county_num  \
0  2004    AL  Autauga County     01001    15196     4758       20081   
1  2008    AL  Autauga County     01001    17403     6093       23641   
2  2004    AL  Baldwin County     01003    52971    15599       69320   
3  2008    AL  Baldwin County     01003    61271    19386       81413   
4  2004    AL  Barbour County     01005     5899     4832       10777   

   dem_margin     vap   turnout  
0   -0.519795   35315  0.568625  
1   -0.478406   35315  0.669432  
2   -0.539123  128945  0.537594  
3   -0.514476  128945  0.631378  
4   -0.099007   22285  0.483599

In [6]:
# Data quality checks:

# voting2_df.sort_values(by='turnout')  #by='dem_margin'

# voting2_df[(voting2_df['turnout'] > 1) & (voting2_df['year'] == 2004)]

# year	state	county_name	fips_code	rep_num	dem_num	county_num	dem_margin	vap	turnout
# 484	2004	CO	Hinsdale County	08053	355	236	602	-0.197674	490	1.228571
# 2254	2004	LA	Saint Bernard Parish	22087	19597	9956	29838	-0.323111	27540	1.083442
# 3254	2004	NE	Arthur County	31005	240	24	266	-0.812030	235	1.131915
# 3554	2004	NM	Harding County	35021	380	259	644	-0.187888	505	1.275248
# 4294	2004	OK	McIntosh County	40091	9946	6933	16879	-0.178506	15420	1.094617
# 4296	2004	OK	Major County	40093	10041	3742	13783	-0.457012	5405	2.550046
# 4298	2004	OK	Marshall County	40095	7472	3684	11156	-0.339548	10340	1.078917

# voting2_df[(voting2_df['turnout'] > 1) & (voting2_df['year'] == 2008)]
# Possible this will affect RI results, 168,000 will be filtered out of a population of 1 million
# We'll see how it works.

# 	year	state	county_name	fips_code	rep_num	dem_num	county_num	dem_margin	vap	turnout
# 485	2008	CO	Hinsdale County	08053	344	240	599	-0.173623	490	1.222449
# 1427	2008	IN	LaPorte County	18091	17918	28258	208757	0.049531	82565	2.528396
# 3193	2008	MT	McCone County	30055	2822	1607	4567	-0.266039	1330	3.433835
# 3255	2008	NE	Arthur County	31005	217	39	263	-0.676806	235	1.119149
# 3555	2008	NM	Harding County	35021	358	260	626	-0.156550	505	1.239604
# 4573	2008	RI	Washington County	44009	49810	116156	168633	0.393434	98525	1.711576
# 4639	2008	SC	McCormick County	45065	5416	9608	15174	0.276262	8435	1.798933
# 4907	2008	TN	McNairy County	47109	23290	20209	43846	-0.070269	19350	2.265943
# 4913	2008	TN	Marion County	47115	20288	13058	33767	-0.214114	21455	1.573852

# print(
# voting2_df[voting2_df['turnout'] > 1]['county_num'].count(),
# voting2_df[voting2_df['turnout'] > 1]['county_num'].sum(),
# voting2_df[(voting2_df['turnout'] > 1) & 
#             (voting2_df['year'] == 2004)]['county_num'].sum(),
# voting2_df[(voting2_df['turnout'] > 1) & 
#             (voting2_df['year'] == 2008)]['county_num'].sum()
# )
# So there are 16 counties with turnout over 1
# and they make up 549400 votes. 
# with 476232 coming from 2008, and 73168 from 2004
# 16 549400 73168 476232

# These counties have negative turnout. . . 
# 1039	2008	HI	Kalawao County	15005	-9999	-9999	-9999	-0.000000	80	-124.987500
# 1038	2004	HI	Kalawao County	15005	-9999	-9999	-9999	-0.000000	80	-124.987500


# MO switches sides in results:
# voting2_df[(voting2_df['state'] == 'MO') & 
#            (voting2_df['year'] == 2008) &
#            ((voting2_df['turnout'] > 1) | (voting2_df['dem_margin'] > 1) | (voting2_df['dem_margin'] < -1)) ]

# This is the cause:
# 	year	state	county_name	fips_code	rep_num	dem_num	county_num	dem_margin	vap	turnout
# 3013	2008	MO	Laclede County	29105	10875	5218	2024	-2.79496	25950	0.077996
# Fixed above.  


# voting2_df['fips_code'].drop_duplicates().count()
# 3154 counties, other is 3112, so probably alaska is main difference? but that's only ~ 10-20
# Note, without AK = 3114, close enough.  2 missing ones are from other dataset that I removed.  I could look
# those counties up, then these would match exactly.  pretty good.  

# voting2_df.sort_values(by='dem_margin')

# -200% margin here.
# 4660	2008	MO	Laclede County	29105	10875	5218	2024	-2.794960
# vs https://en.wikipedia.org/wiki/United_States_presidential_election_in_Missouri,_2008
# Laclede	32.4%	5,218	67.6%	10,875	16,093
# Issue is that county num is 2023 when it should be around 16093

# https://en.wikipedia.org/wiki/Oglala_Lakota_County,_South_Dakota
# Oglala Lakota County, known as Shannon County until May 2015, is a county located 
# in the U.S. state of South Dakota. The population was 13,586 at the 2010 census. 
# Oglala Lakota County does not have its own county seat. Wikipedia
# have the same fips code though, so should be fine.  

# 2004
# print('2004 Results:')
# print('Dem: ', voting2_df[(voting2_df['year'] == 2004)]['dem_num'].sum())
# print('Rep: ', voting2_df[(voting2_df['year'] == 2004)]['rep_num'].sum())
# print('Total: ', voting2_df[(voting2_df['year'] == 2004)]['county_num'].sum())

# # 2008
# print('2008 Results:')
# print('Dem: ', voting2_df[(voting2_df['year'] == 2008)]['dem_num'].sum())
# print('Rep: ', voting2_df[(voting2_df['year'] == 2008)]['rep_num'].sum())
# print('Total: ', voting2_df[(voting2_df['year'] == 2008)]['county_num'].sum())

# 2004 Results:
# Dem:  59093576
# Rep:  62131489
# Total:  122459692
# 2008 Results:
# Dem:  69547395
# Rep:  59959177
# Total:  131502027

# Leip atlas:
# 2004
# http://uselectionatlas.org/RESULTS/
# 	George W. Bush	Richard Cheney	Republican	62,039,572	50.73%	286	53.2%
# 	John Kerry	John Edwards	Democratic	59,027,115	48.26%	251	46.7%
# Total	122,303,590		538	

# 2008
# 	Barack H. Obama	Joseph R. Biden, Jr.	Democratic	69,499,428	52.86%	365	67.8%
# 	John S. McCain, III	Sarah H. Palin	Republican	59,950,323	45.60%	173	32.2%
# Total	131,473,705		538	

# Ok, so pretty close/good enough.  

In [7]:
# Source: https://en.wikipedia.org/wiki/Electoral_College_(United_States)#Chronological_table
electoral_df = pd.read_csv('./data/state_electoral_votes.csv') #sep="\t"

# Electoral points in 2004 = 2008, 2012=2016=2020
electoral_df['votes_2016'] = electoral_df['votes_2012']
electoral_df['votes_2004'] = electoral_df['votes_2008']
electoral_df['state'] = electoral_df['state'].apply(lambda x: state_abbrev[x])

years = [2004, 2008, 2012, 2016]
electoral_dfs = []

#Stack by year
for year in years:
    cols = 'state votes_{0}'.format(year)
    cols = cols.split(' ')
    temp_df = electoral_df[cols].copy()
    temp_df['year'] = year
    temp_df.rename(columns={'votes_{}'.format(year): 'state_electoral_votes'}, inplace=True)
    electoral_dfs.append(temp_df)

electoral_df = pd.concat(electoral_dfs)

electoral_df.head()

state  state_electoral_votes  year
0    AL                      9  2004
1    AK                      3  2004
2    AZ                     10  2004
3    AR                      6  2004
4    CA                     55  2004

In [8]:
# Manually input Alaska statewide data because they don't use counties
# http://uselectionatlas.org/RESULTS/compare.php?year=2016&fips=2&f=0&off=0&elect=0&type=state

columns = ['fips_code', 'county_num', 'dem_num', 'rep_num', 
            'year', 'state', 'county_name', 'vap']

ak = [['02', '02', '02', '02'], [312598, 326197, 300495, 318608], [111025, 123594, 122640, 116454], 
      [190889, 193841, 164676, 163387], [2004, 2008, 2012, 2016], ['AK', 'AK', 'AK', 'AK'],
      ['Alaska', 'Alaska', 'Alaska', 'Alaska'], [550189, 550189, 550189, 550189]]

alaska_df = pd.DataFrame(dict(zip(columns, ak)))

alaska_df['dem_margin'] =  (alaska_df['dem_num'] - alaska_df['rep_num'])/ alaska_df['county_num']
alaska_df['turnout'] = alaska_df['county_num']/ alaska_df['vap']

alaska_df


county_name  county_num  dem_num fips_code  rep_num state     vap  year  \
0      Alaska      312598   111025        02   190889    AK  550189  2004   
1      Alaska      326197   123594        02   193841    AK  550189  2008   
2      Alaska      300495   122640        02   164676    AK  550189  2012   
3      Alaska      318608   116454        02   163387    AK  550189  2016   

   dem_margin   turnout  
0   -0.255485  0.568165  
1   -0.215351  0.592882  
2   -0.139889  0.546167  
3   -0.147306  0.579088

In [9]:
# Combine all dataframes together
turnout_df = pd.concat([voting1_df, voting2_df, alaska_df])

#Add in electoral college points for each year
turnout_df = pd.merge(turnout_df, electoral_df, on=('state', 'year'))

turnout_df.head()

county_name  county_num  dem_margin  dem_num fips_code  rep_num state  \
0  Autauga County       23973   -0.459517     6363     01001    17379    AL   
1  Baldwin County       85491   -0.556690    18424     01003    66016    AL   
2  Barbour County       11517    0.031432     5912     01005     5550    AL   
3     Bibb County        8420   -0.466746     2202     01007     6132    AL   
4   Blount County       24060   -0.739277     2970     01009    20757    AL   

   turnout     vap  year  state_electoral_votes  
0   0.5865   40874  2012                      9  
1   0.5799  147416  2012                      9  
2   0.5398   21334  2012                      9  
3   0.4731   17796  2012                      9  
4   0.5484   43876  2012                      9

In [10]:
#Data output script for interactive visualization, filters out abnormal values  
out_df = turnout_df[(turnout_df['dem_margin'] <= 1.0) &
                    (turnout_df['dem_margin'] >= -1.0) &
                    (turnout_df['turnout'] >= 0.0) &
                    (turnout_df['turnout'] <= 1.0)]

# out_df = out_df[['county', 'state', 'county_num', 'turnout', 'num_rep', 'num_dem', 'num_state_dem',
#                  'num_state_rep','num_state', 'year', 'vap', 'fips_code', 'state_electoral_votes']]

out_df = out_df.round(decimals=4)

out_df.to_csv('./US_County_Level_Presidential_Results_04-16.csv', index=False)

In [11]:
state_df = turnout_df[(turnout_df['dem_margin'] <= 1.0) &
                    (turnout_df['dem_margin'] >= -1.0) &
                    (turnout_df['turnout'] >= 0.0) &
                    (turnout_df['turnout'] <= 1.0)]

state_df = state_df.groupby(['state','year']).agg(
    {'county_num':'sum', 'dem_num':'sum','rep_num':'sum',
     'vap':'sum', 'state_electoral_votes':'max'})

state_df = state_df.round(decimals=4)

state_df = state_df.reset_index()
state_df.rename(columns={'county_num':'state_num'}, inplace=True)
state_df

state  year  dem_num  rep_num  state_electoral_votes       vap  state_num
0      AK  2004   111025   190889                      3    550189     312598
1      AK  2008   123594   193841                      3    550189     326197
2      AK  2012   122640   164676                      3    550189     300495
3      AK  2016   116454   163387                      3    550189     318608
4      AL  2004   693933  1176394                      9   3430120    1883415
5      AL  2008   813479  1266546                      9   3430120    2099819
6      AL  2012   795696  1255925                      9   3699846    2074338
7      AL  2016   729547  1318255                      9   3771201    2123372
8      AR  2004   469953   572898                      6   2067620    1054945
9      AR  2008   422310   638017                      6   2067620    1086617
10     AR  2012   394409   647744                      6   2239501    1069468
11     AR  2016   380494   684872                      6   2285967    1130635
12     AZ  2004   893524  1104462                     10   4087540    2012753
13     AZ  2008  1034707  1230111                     10   4087540    2293475
14     AZ  2012  1025232  1233654                     11   4939258    2306559
15     AZ  2016  1161167  1252401                     11   5302793    2554240
16     CA  2004  6745485  5509826                     55  21942935   12419372
17     CA  2008  8274473  5011781                     55  21942935   13561900
18     CA  2012  7854285  4839958                     55  28854869   13055815
19     CA  2016  8753788  4483810                     55  30401787   14236960
20     CO  2004  1001469  1100901                      9   3363945    2129702
21     CO  2008  1288336  1073245                      9   3363945    2400762
22     CO  2012  1322811  1184899                      9   3957271    2571196
23     CO  2016  1338633  1202140                      9   4292250    2779568
24     CT  2004   857488   693826                      7   2460485    1578769
25     CT  2008   997772   629428                      7   2460485    1646792
26     CT  2012   905109   634899                      7   2798474    1558075
27     CT  2016   895752   673215                      7   2833744    1634650
28     DC  2004   202970    21256                      3    430110     227586
29     DC  2008   245800    17367                      3    430110     265853
..    ...   ...      ...      ...                    ...       ...        ...
174    TX  2012  3308124  4569843                     38  19104772    7999657
175    TX  2016  3877868  4685047                     38  20672745    8969226
176    UT  2004   241199   663742                      5   1703980     927844
177    UT  2008   327670   596030                      5   1703980     952370
178    UT  2012   251813   740600                      6   1967463    1020861
179    UT  2016   310674   515211                      6   2128191    1131317
180    VA  2004  1454742  1716959                     13   5498720    3198852
181    VA  2008  1959469  1725005                     13   5498720    3723197
182    VA  2012  1970595  1820995                     13   6330729    3851684
183    VA  2016  1981473  1769443                     13   6566663    3982752
184    VT  2004   184067   121180                      3    479705     312309
185    VT  2008   219262    98974                      3    479705     325046
186    VT  2012   199239    92698                      3    501575     299290
187    VT  2016   178573    95369                      3    507112     313080
188    WA  2004  1510201  1304894                     11   4533610    2859084
189    WA  2008  1750848  1229216                     11   4533610    3036878
190    WA  2012  1755396  1290670                     12   5309540    3145958
191    WA  2016  1742718  1221747                     12   5657776    3294874
192    WI  2004  1489504  1478120                     10   4155115    2997004
193    WI  2008  1676761  

In [70]:
#state_df['turnout'] = state_df['state_num']/state_df['vap']
#list(state_df)
#state_df = state_df.round(decimals=4)
state_df.to_csv('./US_State_Level_Presidential_Results_04-16.csv', index=False)


In [12]:
national_df = turnout_df[(turnout_df['dem_margin'] <= 1.0) &
                    (turnout_df['dem_margin'] >= -1.0) &
                    (turnout_df['turnout'] >= 0.0) &
                    (turnout_df['turnout'] <= 1.0)]

national_df = national_df.groupby(['year']).agg(
    {'county_num':'sum', 'dem_num':'sum','rep_num':'sum',
     'vap':'sum'})

national_df.rename(columns={'county_num':'num_nation'}, inplace=True)

national_df = national_df.reset_index()

national_df

year   dem_num   rep_num        vap  num_nation
0  2004  59006163  61992079  208206569   122217180
1  2008  69367959  59848713  208033614   131049863
2  2012  65912769  60929447  240396331   129254383
3  2016  65805766  62941275  250285285   136670626

# Demographic Data

# ANES 2016 Race Data

In [13]:
## Calculate 2016 Democratic Margin By Race
## source: ANES 2016 time series study:
## http://www.electionstudies.org/studypages/anes_timeseries_2016/anes_timeseries_2016.htm

anes_df = pd.read_csv('./data/anes/anes_timeseries_2016/anes_timeseries_2016_rawdata.txt',
                      sep='|')

columns = ['V160102', 'V161310x', 'V162058x']
labels = ['weight_postfull', 'group', 'party']

anes_df = anes_df[columns]
anes_df.rename(columns=dict(zip(columns, labels)), inplace=True)

anes_df = anes_df[anes_df['party'].isin([10,11,12])]  #Filter out nonvoters

group_df = anes_df.groupby(['group', 'party']).agg({'weight_postfull':'sum'})
group_df = group_df.unstack(level=-1)
group_df.columns = group_df.columns.droplevel(0)
group_df = group_df.rename_axis(None, axis=1)

group_df.loc[6] = group_df.loc[[3,4,6,-9]].sum()  #Include nonresponders in other, -9
rows = [1,2,5,6]
races = ['Non-Hispanic White', 'Non-Hispanic Black', 'Hispanic', 'Other']
group_df.rename(index=dict(zip(rows,races)), inplace=True)
group_df = group_df.loc[races] 

group_df.reset_index(inplace=True)
group_df['total'] = group_df[[10,11,12]].sum(axis=1)
group_df.rename(columns={10:'d',11:'r',12:'o'}, inplace=True)

group_df['dem_frac'] = group_df['d']/group_df['total']
group_df['rep_frac'] = group_df['r']/group_df['total']
group_df['other_frac'] = group_df['o']/group_df['total']
#group_df['dem_margin'] = group_df['dem_frac'] - group_df['rep_frac']

group_df['year'] = 2016

exp_df = group_df.copy()
group_df = group_df[['group','year','dem_frac', 'rep_frac']]
group_df


/Users/psthomas/miniconda2/envs/datascience3/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (790,1129,1131) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


group  year  dem_frac  rep_frac
0  Non-Hispanic White  2016  0.393836  0.538035
1  Non-Hispanic Black  2016  0.899861  0.063772
2            Hispanic  2016  0.686294  0.238819
3               Other  2016  0.573222  0.314705

In [169]:
exp_df = exp_df.merge(national_df, on='year')
exp_df['electorate_frac'] = exp_df['total'] / exp_df['total'].sum()

exp_df['altdemnum'] = exp_df['electorate_frac']*exp_df['dem_frac']*exp_df['num_nation']
exp_df['altrepnum'] = exp_df['electorate_frac']*exp_df['rep_frac']*exp_df['num_nation']

print(exp_df['altdemnum'].sum()/exp_df['num_nation'].mean())
print(exp_df['altrepnum'].sum()/exp_df['num_nation'].mean())

print(exp_df['dem_num'].mean()/exp_df['num_nation'].mean())
print(exp_df['rep_num'].mean()/exp_df['num_nation'].mean())
# exp_df

0.49001285898
0.44153235198
0.481491655712
0.460532572669


The results of the demographics extrapolated don't agree with the 
actual cumulative results:  
Extrapolated:  
D: 0.49001285898  
R: 0.44153235198  
Actual:  
D: 0.481491655712  
R: 0.460532572669  

This is a commonly known problem: 

https://www.americanprogress.org/issues/democracy/reports/2017/11/01/441926/voter-trends-in-2016/

> The most vexing issue for electoral analysts exploring these voting trends is determining which, if any, of the existing data sources provide the best and most reliable information on vote composition, turnout, and support rates. For this project, we developed original turnout and support estimates by combining a multitude of publicly available data sources including the American Communities Survey (ACS), the November supplement of the Current Population Survey (CPS), the American National Election Study (ANES), the Cooperative Congressional Election Survey (CCES), our own post-election polling, and voter files from several states.
We used this approach to help address what we believe are systematic problems with some of the most widely available and most frequently cited pieces of data about elections—mainly, that some of the most reliable sources of data we have on demographics do not fit well together with the best data we have on turnout rates, leading to results that vary from the actual levels of turnout seen on Election Day. Furthermore, if we combine those data with the best data we have on vote choice, we get election results that do not line up with reality. This is not due to any one source of information being particularly biased; rather, each particular source has points of weakness. To overcome this, we created a new method for combining these data in ways that fit with known outcomes.

## ANES and CPS 2004-2012 Race Data

In [14]:
#Sources:
#http://www.electproject.org/home/voter-turnout/demographics
#https://docs.google.com/spreadsheets/d/1l5fpK7ysQhQbZPv9hnZ_-PO1J1zBVPXSSQjNejTXecY/edit#gid=0
#Below includes both corrected and uncorrected data, as data corrected for turnout response
#bias doesn't result in correct overall D/R voter percentages.
#Neither corrected or uncorrected result in right count.  
#These turnouts are for citizen voting age population:

# Corrected USEP turnout data:
# race_turnout = StringIO("""
# Turnout Rate	2016	2014	2012	2010	2008	2006	2004
# Non-Hispanic White	64.7%	40.8%	61.8%	45.0%	65.2%	44.7%	64.3%
# Non-Hispanic Black	59.9%	36.4%	67.4%	41.6%	69.1%	36.6%	61.4%
# Hispanic	44.9%	21.1%	43.1%	26.6%	46.5%	25.5%	42.9%
# Other	46.3%	24.5%	45.4%	30.7%	48.0%	28.5%	44.9%
# """)

#Uncorrected USEP turnout data:
race_turnout = StringIO("""
Turnout Rate	2016	2014	2012	2010	2008	2006	2004
Non-Hispanic White	65.3%	45.8%	64.1%	48.6%	66.1%	51.6%	67.2%
Non-Hispanic Black	59.3%	40.4%	66.4%	43.8%	65.2%	41.2%	60.3%
Hispanic	47.6%	27.0%	48.0%	31.2%	49.9%	32.3%	47.2%
Other	49.0%	29.5%	49.0%	34.1%	49.5%	33.4%	48.0%
""")

# Corrected USEP share of electorate data:
# race_share = StringIO("""
# Share of Electorate	2016	2014	2012	2010	2008	2006	2004
# Non-Hispanic White	73.6%	76.9%	74.1%	77.9%	76.6%	81.0%	79.5%
# Non-Hispanic Black	12.3%	11.9%	13.2%	11.4%	12.3%	10.0%	11.1%
# Hispanic	9.1%	7.0%	8.3%	6.8%	7.3%	5.6%	5.9%
# Other	5.0%	4.2%	4.4%	3.9%	3.8%	3.4%	3.6%
# """)

#Try uncorrected USEP data:
race_share = StringIO("""
Share of Electorate	2016	2014	2012	2010	2008	2006	2004
Non-Hispanic White	73.3%	76.3%	73.7%	77.5%	76.3%	80.4%	79.2%
Non-Hispanic Black	12.4%	12.1%	13.4%	11.7%	12.4%	10.3%	11.2%
Hispanic	9.2%	7.3%	8.4%	6.9%	7.4%	5.8%	6.0%
Other	5.0%	4.3%	4.5%	3.9%	3.9%	3.4%	3.6%
""")

#Source:
#http://www.electionstudies.org/nesguide/text/t9a_1_1.txt
race_demmargin = StringIO("""
percent_of_group,1948,1952,1954,1956,1958,1960,1962,1964,1966,1968,1970,1972,1974,1976,1978,1980,1982,1984,1986,1988,1990,1992,1994,1996,1998,2000,2002,2004,2008,2012
Non-Hispanic White,53,40,,39,,48,,65,,41,,30,,46,,36,,36,,39,,52,,51,,46,,42,44,43,
Non-Hispanic Black,65*,80,,64*,,71,,100,,97,,87,,95,,93,,91,,92,,95,,99,,92,,89,100,97,
Hispanic,,,,,,,,,,40*,,50*,,86*,,58*,,52,,65,,65,,83,,59,,61,73,73,
Other,,,,,,,,,,33*,,70*,,43*,,25*,,50*,,83*,,44*,,54*,,57,,36*,82,61
""")

race_repmargin = StringIO("""
percent_of_group,2004,2008,2012
Non-Hispanic White,58,56,57 
Non-Hispanic Black,11,0,3 
Hispanic,39,27,27 
Other,64,18,39 
""")

raceturnout_df = pd.read_csv(race_turnout, sep='\t') #.T
raceturnout_df = pd.melt(raceturnout_df, id_vars=["Turnout Rate"], 
                  var_name="year", value_name="turnout")
raceturnout_df.rename(columns={'Turnout Rate':'group'}, inplace=True)

raceshare_df = pd.read_csv(race_share, sep='\t')
raceshare_df = pd.melt(raceshare_df, id_vars=["Share of Electorate"], 
                  var_name="year", value_name="electorate_frac")
raceshare_df.rename(columns={'Share of Electorate':'group'}, inplace=True)

racedemmargin_df = pd.read_csv(race_demmargin, sep=',',index_col=False)
racedemmargin_df = racedemmargin_df[['percent_of_group', '2004', '2008', '2012']]
racedemmargin_df = pd.melt(racedemmargin_df, id_vars=["percent_of_group"], 
                  var_name="year", value_name="dem_frac")
racedemmargin_df.rename(columns={'percent_of_group':'group'}, inplace=True)

racerepmargin_df = pd.read_csv(race_repmargin, sep=',',index_col=False)
racerepmargin_df = racerepmargin_df[['percent_of_group', '2004', '2008', '2012']]
racerepmargin_df = pd.melt(racerepmargin_df, id_vars=["percent_of_group"], 
                  var_name="year", value_name="rep_frac")
racerepmargin_df.rename(columns={'percent_of_group':'group'}, inplace=True)

race_df = raceturnout_df.merge(raceshare_df, on=['group','year'], how='inner')
race_df = race_df.merge(racedemmargin_df, on=['group','year'], how='inner')
race_df = race_df.merge(racerepmargin_df, on=['group','year'], how='inner')
race_df = race_df.replace({'\%': '', '\*':''}, regex=True)
race_df = race_df.apply(pd.to_numeric, errors='ignore', axis=0)
race_df[['turnout', 'electorate_frac',
         'dem_frac', 'rep_frac']] = race_df[['turnout', 'electorate_frac', 
                                             'dem_frac','rep_frac']] / 100

#Add in 2016 data from ANES
int_df = raceturnout_df.merge(raceshare_df, on=['group','year'], how='inner')
int_df = int_df.replace({'\%': '', '\*':''}, regex=True)
int_df = int_df.apply(pd.to_numeric, errors='ignore', axis=0)
int_df = int_df.merge(group_df, on=['group','year'], how='inner')
int_df['electorate_frac'] = int_df['electorate_frac']/100
int_df['turnout'] = int_df['turnout']/100
race_df = pd.concat([race_df, int_df])

electorate_df = national_df[['year','num_nation']]
race_df = race_df.merge(electorate_df, on='year', how='left')

race_df['dem_margin'] = race_df['dem_frac'] - race_df['rep_frac']
#race_df['num_group'] = race_df['electorate_frac']*race_df['num_nation']
#race_df['num_dem'] = race_df['dem_frac']*race_df['num_group']  #Calculate when loaded into JS
#race_df['num_rep'] = race_df['rep_frac']*race_df['num_group']  #Calculate when loaded into JS
#e.g. race_df['num_dem'] = race_df['dem_frac']*race_df['num_nation']*race_df['electorate_frac']

race_df['demographic'] = 'race'
order = ['demographic','group', 'year','dem_frac','rep_frac','electorate_frac', 'num_nation', 'dem_margin', 'turnout']
race_df = race_df[order]

race_df

demographic               group  year  dem_frac  rep_frac  electorate_frac  \
0         race  Non-Hispanic White  2012  0.430000  0.570000            0.737   
1         race  Non-Hispanic Black  2012  0.970000  0.030000            0.134   
2         race            Hispanic  2012  0.730000  0.270000            0.084   
3         race               Other  2012  0.610000  0.390000            0.045   
4         race  Non-Hispanic White  2008  0.440000  0.560000            0.763   
5         race  Non-Hispanic Black  2008  1.000000  0.000000            0.124   
6         race            Hispanic  2008  0.730000  0.270000            0.074   
7         race               Other  2008  0.820000  0.180000            0.039   
8         race  Non-Hispanic White  2004  0.420000  0.580000            0.792   
9         race  Non-Hispanic Black  2004  0.890000  0.110000            0.112   
10        race            Hispanic  2004  0.610000  0.390000            0.060   
11        race               Other  2004  0.360000  0.640000            0.036   
12        race  Non-Hispanic White  2016  0.393836  0.538035            0.733   
13        race  Non-Hispanic Black  2016  0.899861  0.063772            0.124   
14        race            Hispanic  2016  0.686294  0.238819            0.092   
15        race               Other  2016  0.573222  0.314705            0.050   

    num_nation  dem_margin  turnout  
0    129254383   -0.140000    0.641  
1    129254383    0.940000    0.664  
2    129254383    0.460000    0.480  
3    129254383    0.220000    0.490  
4    131049863   -0.120000    0.661  
5    131049863    1.000000    0.652  
6    131049863    0.460000    0.499  
7    131049863    0.640000    0.495  
8    122217180   -0.160000    0.672  
9    122217180    0.780000    0.603  
10   122217180    0.220000    0.472  
11   122217180   -0.280000    0.480  
12   136670626   -0.144198    0.653  
13   136670626    0.836089    0.593  
14   136670626    0.447475    0.476  
15   136670626    0.258517    0.490

In [15]:
#out_df = out_df.round(decimals=4), don't round, numbers are huge
#race_df.to_csv('./US_Race_Presidential_Results_04-16.csv', index=False)

# ANES and CPS 2004-2012, 2016 Gender Data

In [16]:
## Calculate 2016 Democratic Margin By Gender
## source: ANES time series cumulative:
## http://www.electionstudies.org/studypages/anes_timeseries_cdf/anes_timeseries_cdf.htm
anes2016_df = pd.read_csv('./data/anes/anes_timeseries_2016/anes_timeseries_2016_rawdata.txt',
                      sep='|')

columns = ['V160102', 'V161342', 'V162058x']
labels = ['weight_postfull', 'group', 'party']

anes2016_df = anes2016_df[columns]
anes2016_df.rename(columns=dict(zip(columns, labels)), inplace=True)

anes2016_df = anes2016_df[anes2016_df['party'].isin([10,11,12])]  #Filter out nonvoters

gender2016_df = anes2016_df.groupby(['group', 'party']).agg({'weight_postfull':'sum'})
gender2016_df = gender2016_df.unstack(level=-1)
gender2016_df.columns = gender2016_df.columns.droplevel(0)
gender2016_df = gender2016_df.rename_axis(None, axis=1)

rows = [1,2,3,-9]
races = ['Male', 'Female', 'Other', 'Nonresponse']
gender2016_df.rename(index=dict(zip(rows,races)), inplace=True)
gender2016_df = gender2016_df.loc[races] 

gender2016_df.reset_index(inplace=True)
gender2016_df['total'] = gender2016_df[[10,11,12]].sum(axis=1)
gender2016_df.rename(columns={10:'d',11:'r',12:'o'}, inplace=True)

gender2016_df['dem_frac'] = gender2016_df['d']/gender2016_df['total']
gender2016_df['rep_frac'] = gender2016_df['r']/gender2016_df['total']
gender2016_df['other_frac'] = gender2016_df['o']/gender2016_df['total']

#0 R votes among 'other'
gender2016_df.fillna(0, inplace=True)

#Create electorate fraction
gender2016_df['electorate_frac'] = gender2016_df['total'] / gender2016_df['total'].sum()

gender2016_df['year'] = 2016
gender2016_df['demographic'] = 'gender'
gender2016_df = gender2016_df[['demographic', 'group','year','dem_frac', 'rep_frac',
                               'other_frac','electorate_frac']]

gender2016_df = gender2016_df[gender2016_df['group'].isin(['Male','Female'])]
gender2016_df['group'].replace({'Male':'male','Female':'female'}, inplace=True)

gender2016_df

/Users/psthomas/miniconda2/envs/datascience3/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (790,1129,1131) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


demographic   group  year  dem_frac  rep_frac  other_frac  electorate_frac
0      gender    male  2016  0.454789  0.469016    0.076195         0.462845
1      gender  female  2016  0.520975  0.419133    0.059892         0.524759

In [17]:
baseanes2004_df = pd.read_csv('./data/anes/anes_timeseries_cdf/anes_timeseries_cdf_rawdata.txt',
                      sep='|' )  
#, low_memory=False
#error_bad_lines=False, warn_bad_lines=False

columns = ['VCF0004', 'VCF0009z', 'VCF0104', 'VCF0102', 'VCF0110', 'VCF0705', 'VCF0702']
labels = ['year', 'weight_postfull', 'gender', 'age','education', 'party', 'voted']

baseanes2004_df = baseanes2004_df[columns]
baseanes2004_df.rename(columns=dict(zip(columns, labels)), inplace=True)
baseanes2004_df = baseanes2004_df[baseanes2004_df['year'].isin([2004,2008,2012])]

/Users/psthomas/miniconda2/envs/datascience3/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (13,14,15,18,19,22,23,24,25,26,27,28,29,30,31,32,33,34,35,39,40,41,42,43,44,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,234,255,276,297,298,299,300,301,302,303,304,305,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337

In [18]:
## Calculate 2004-2012 Democratic Margin By Gender
## source: ANES time series cumulative:
## http://www.electionstudies.org/studypages/anes_timeseries_cdf/anes_timeseries_cdf.htm
gender2004_df = baseanes2004_df[baseanes2004_df['party'].isin([1,2,3])].copy()  #Filter out nonvoters
gender2004_df.rename(columns={'gender':'group'},inplace=True)
gender2004_df = gender2004_df.groupby(['year', 'group', 'party']).agg({'weight_postfull':'sum'})
gender2004_df = gender2004_df.unstack(level=-1)
gender2004_df.columns = gender2004_df.columns.droplevel(0)
gender2004_df = gender2004_df.rename_axis(None, axis=1)

#Reset index, then just relabel by applying to column, no need to select out
gender2004_df.reset_index(inplace=True)
gender2004_df['group'] = gender2004_df['group'].replace({1:'male',2:'female'})
gender2004_df['total'] = gender2004_df[[1,2,3]].sum(axis=1)
gender2004_df.rename(columns={1:'d',2:'r',3:'o'}, inplace=True)

#Divide the number of respondents for each gender by total gender respondents
#to get fraction of electorate
sum_df = gender2004_df.groupby('year')['total'].sum()
def electorate_frac(row):
    row['electorate_frac'] = row['total']/sum_df[row['year']]
    return row 
gender2004_df = gender2004_df.apply(electorate_frac, axis=1)

gender2004_df['dem_frac'] = gender2004_df['d']/gender2004_df['total']
gender2004_df['rep_frac'] = gender2004_df['r']/gender2004_df['total']
gender2004_df['other_frac'] = gender2004_df['o']/gender2004_df['total']

gender2004_df['demographic'] = 'gender'
gender2004_df = gender2004_df[['group','year','dem_frac', 'rep_frac', 
                               'other_frac', 'electorate_frac', 'demographic']]

gender2004_df

group  year  dem_frac  rep_frac  other_frac  electorate_frac demographic
0    male  2004  0.478701  0.491888    0.029411         0.519864      gender
1  female  2004  0.444623  0.544912    0.010465         0.480136      gender
2    male  2008  0.505027  0.467885    0.027088         0.429771      gender
3  female  2008  0.561853  0.423652    0.014495         0.570229      gender
4    male  2012  0.486533  0.469894    0.043573         0.478998      gender
5  female  2012  0.549398  0.430607    0.019995         0.521002      gender

In [19]:
#Concat 2016 data on, and then merge with num_nation and turnout data:
gender_df = pd.concat([gender2004_df, gender2016_df])

#Add in number in electorate
electorate_df = national_df[['year','num_nation']]
gender_df = gender_df.merge(electorate_df, on='year', how='left')

#2004-2016 Turnout by Gender
#It's possible to calculate this from ANES data, but it is really off.  See cell below
#Source Current Population Survey data:
#https://www.census.gov/data/tables/2004/demo/voting-and-registration/p20-562-rv.html
#https://www.census.gov/data/tables/2008/demo/voting-and-registration/p20-562-rv.html
#https://www.census.gov/data/tables/2012/demo/voting-and-registration/p20-568.html
#https://www.census.gov/data/tables/2016/demo/voting-and-registration/p20-562-rv.html
#All are percent of citizens voting
gender_turnout = StringIO("""
year,group,turnout
2004,male,62.1
2004,female,65.4
2008,male,61.5
2008,female,65.7
2012,male,59.7
2012,female,63.7
2016,male,59.3
2016,female,63.3
""")
gender_turnout_df = pd.read_csv(gender_turnout, sep=',')
gender_turnout_df = gender_turnout_df.apply(pd.to_numeric, errors='ignore', axis=0)
gender_turnout_df['turnout'] = gender_turnout_df['turnout']/100
gender_df = gender_df.merge(gender_turnout_df, on=['year','group'])

gender_df['dem_margin'] = gender_df['dem_frac'] - gender_df['rep_frac']

order = ['demographic','group', 'year','dem_frac','rep_frac','electorate_frac', 'num_nation', 'dem_margin', 'turnout']

gender_df = gender_df[order]

gender_df

demographic   group  year  dem_frac  rep_frac  electorate_frac  num_nation  \
0      gender    male  2004  0.478701  0.491888         0.519864   122217180   
1      gender  female  2004  0.444623  0.544912         0.480136   122217180   
2      gender    male  2008  0.505027  0.467885         0.429771   131049863   
3      gender  female  2008  0.561853  0.423652         0.570229   131049863   
4      gender    male  2012  0.486533  0.469894         0.478998   129254383   
5      gender  female  2012  0.549398  0.430607         0.521002   129254383   
6      gender    male  2016  0.454789  0.469016         0.462845   136670626   
7      gender  female  2016  0.520975  0.419133         0.524759   136670626   

   dem_margin  turnout  
0   -0.013188    0.621  
1   -0.100289    0.654  
2    0.037142    0.615  
3    0.138201    0.657  
4    0.016639    0.597  
5    0.118791    0.637  
6   -0.014226    0.593  
7    0.101842    0.633

In [20]:
turnout_df = baseanes2004_df.copy()
turnout_df = turnout_df.groupby(['year', 'voted']).agg({'weight_postfull': 'sum'})
turnout_df
#This would put 2008 turnout at 70%
#While this only has it at 62%
#http://www.electproject.org/2008g
#            VEP (all ballots) VEP   VAP
#2004 United States	60.7%	60.1%	55.4%
#2008 United States	62.2%	61.6%	56.9%
#2012 United States	58.6%	58.0%	53.6%
#2016 United States	60.2%	59.3%	54.7%

# 2004 812/(812+248+150) = 67.17%
# 2008 1639/(473+208+1639) = 70.6%
# 2012 4282/(4282+1218+413) = 72.4%
# So it does seem that raw surveys overestimate voter turnout substantially.
# So stick with the adjusted CPS Turnout rates from the elections project,
# and the demographic data from them as well when possible. 

weight_postfull
year voted                 
2004 0             150.9553
     1             248.7554
     2             812.2912
2008 0             208.9927
     1             473.0947
     2            1639.9759
2012 0             413.1968
     1            1218.2133
     2            4282.5888

# Age data from CPS, ANES, 2004-2016

I obtain turnout and share of the electorate data frome CPS, the democratic margin by age from the ANES data,
and the number of national voters from the cumulative county data.  


In [33]:
## Calculate 2016 Democratic Margin By Age Group
## source: ANES time series cumulative:
## http://www.electionstudies.org/studypages/anes_timeseries_cdf/anes_timeseries_cdf.htm

# V161267x
# Label: Item name:
# Question:
# PRE: SUMMARY - Respondent age group
# Not applicable; administrative or derived variable
# Not applicable
# Unweighted Frequencies
# 01. Age group 18-20 
# 02. Age group 21-24 
# 03. Age group 25-29 
# 04. Age group 30-34 
# 05. Age group 35-39 
# 06. Age group 40-44 
# 07. Age group 45-49 
# 08. Age group 50-54 
# 09. Age group 55-59 
# 10. Age group 60-64 
# 11. Age group 65-69 
# 12. Age group 70-74
# 13. Age group 75 or older
# -1. Inapplicable

anes2016_df = pd.read_csv('./data/anes/anes_timeseries_2016/anes_timeseries_2016_rawdata.txt',
                      sep='|')

columns = ['V160102', 'V161267x', 'V162058x']
labels = ['weight_postfull', 'group', 'party']

anes2016_df = anes2016_df[columns]
anes2016_df.rename(columns=dict(zip(columns, labels)), inplace=True)

anes2016_df = anes2016_df[anes2016_df['party'].isin([10,11,12])]  #Filter out nonvoters

age2016_df = anes2016_df.groupby(['group', 'party']).agg({'weight_postfull':'sum'})
age2016_df = age2016_df.unstack(level=-1)
age2016_df.columns = age2016_df.columns.droplevel(0)
age2016_df = age2016_df.rename_axis(None, axis=1)

# pairs = [[1,2],[3,4],[5,6],[7,8],[9,10],[11,12]]
# labels = {1: "17 - 24", 2: "25 - 34", 3: "35 - 44", 4: "45 - 54", 5: "55 - 64", 6: "65 - 74", 7: "75 +"}

pairs = [[1,2,3],[4,5,6],[7,8,9],[10,11,12]]
labels = {1: "17 - 29", 2: "30 - 44", 3: "45 - 59", 4: "60 +"}

for i, pair in enumerate(pairs):
    age2016_df.loc[labels[i+1]] = age2016_df.loc[pair].sum()
    
age2016_df = age2016_df.loc[labels.values()]


age2016_df.reset_index(inplace=True)
age2016_df['total'] = age2016_df[[10,11,12]].sum(axis=1)
age2016_df.rename(columns={10:'d',11:'r',12:'o'}, inplace=True)

age2016_df['dem_frac'] = age2016_df['d']/age2016_df['total']
age2016_df['rep_frac'] = age2016_df['r']/age2016_df['total']
age2016_df['other_frac'] = age2016_df['o']/age2016_df['total']

#0 R votes among 'other'
age2016_df.fillna(0, inplace=True)

#Create electorate fraction
age2016_df['electorate_frac'] = age2016_df['total'] / age2016_df['total'].sum()


age2016_df['year'] = 2016
age2016_df['demographic'] = 'age'


age2016_df = age2016_df[['demographic', 'group','year','dem_frac', 'rep_frac',
                               'other_frac','electorate_frac']]


age2016_df

/Users/psthomas/miniconda2/envs/datascience3/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (790,1129,1131) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


demographic    group  year  dem_frac  rep_frac  other_frac  electorate_frac
0         age  17 - 29  2016  0.548326  0.351484    0.100191         0.173320
1         age  30 - 44  2016  0.555207  0.345984    0.098809         0.245604
2         age  45 - 59  2016  0.446847  0.479965    0.073187         0.338233
3         age     60 +  2016  0.475273  0.494703    0.030024         0.242843

In [37]:
## Calculate 2004-2012 Democratic Margin By Gender
## source: ANES time series cumulative:
## http://www.electionstudies.org/studypages/anes_timeseries_cdf/anes_timeseries_cdf.htm
# VCF0102

# DEMOGRAPHICS: Respondent - Age Group

# QUESTION:
# ---------
# 1964-1976:  What is your date of birth?
# 1978-1982:  What is the month and year of your birth?
# 1984-LATER: What is the month, day and year of your birth?

# VALID_CODES:
# ------------
# 1.  17 - 24
# 2.  25 - 34
# 3.  35 - 44
# 4.  45 - 54
# 5.  55 - 64
# 6.  65 - 74
# 7.  75 - 99 and over (except 1954)
# MISSING_CODES:
# --------------
# 0.  NA; DK; RF; no Pre IW
# INAP. Inap. question not used

age2004_df = baseanes2004_df[baseanes2004_df['party'].isin([1,2,3])].copy()  #Filter out nonvoters
age2004_df.rename(columns={'age':'group'},inplace=True)
age2004_df = age2004_df.groupby(['year', 'group', 'party']).agg({'weight_postfull':'sum'})
age2004_df = age2004_df.unstack(level=-1)
age2004_df.columns = age2004_df.columns.droplevel(0)
age2004_df = age2004_df.rename_axis(None, axis=1)

age2004_df.fillna(0, inplace=True)
age2004_df.reset_index(inplace=True)

#Interpolate to get correct age ranges
labels = {1: "17 - 29", 2: "30 - 44", 3: "45 - 59", 4: "60 +"}
age2004_df.loc["17 - 29"] = age2004_df.loc[1] + 0.5*age2004_df.loc[2]
age2004_df.loc["30 - 44"] = 0.5*age2004_df.loc[2] + age2004_df.loc[3]
age2004_df.loc["45 - 59"] = age2004_df.loc[4] + 0.5*age2004_df.loc[5]
age2004_df.loc["60 +"] = 0.5*age2004_df.loc[5] + age2004_df.loc[6] + age2004_df.loc[7]

age2004_df['group'] = age2004_df['group'].replace(labels)
age2004_df = age2004_df[age2004_df['group'].isin(labels.values())]
age2004_df['total'] = age2004_df[[1,2,3]].sum(axis=1)
age2004_df.rename(columns={1:'d',2:'r',3:'o'}, inplace=True)

#Divide the number of respondents for each gender by total gender respondents
#to get fraction of electorate
sum_df = age2004_df.groupby('year')['total'].sum()
def electorate_frac(row):
    row['electorate_frac'] = row['total']/sum_df[row['year']]
    return row 
age2004_df = age2004_df.apply(electorate_frac, axis=1)

age2004_df['dem_frac'] = age2004_df['d']/age2004_df['total']
age2004_df['rep_frac'] = age2004_df['r']/age2004_df['total']
age2004_df['other_frac'] = age2004_df['o']/age2004_df['total']

age2004_df['demographic'] = 'age'
age2004_df = age2004_df[['demographic','group','year','dem_frac', 'rep_frac', 
                               'other_frac', 'electorate_frac']]

age2004_df['year'] = age2004_df['year'].astype(int)

age2004_df

demographic    group  year  dem_frac  rep_frac  other_frac  electorate_frac
0          age  17 - 29  2004  0.429017  0.412320    0.158662         0.123634
1          age  30 - 44  2004  0.581622  0.373218    0.045159         0.255757
2          age  45 - 59  2004  0.420608  0.579392    0.000000         0.322568
3          age     60 +  2004  0.445143  0.554857    0.000000         0.298042
8          age  17 - 29  2008  0.671394  0.325719    0.002887         0.141188
9          age  30 - 44  2008  0.629155  0.350683    0.020162         0.268001
10         age  45 - 59  2008  0.562251  0.423448    0.014301         0.250993
11         age     60 +  2008  0.504923  0.465264    0.029813         0.339818
16         age  17 - 29  2012  0.625693  0.338356    0.035951         0.139413
17         age  30 - 44  2012  0.571377  0.368016    0.060607         0.264247
18         age  45 - 59  2012  0.512499  0.447605    0.039897         0.295920
19         age     60 +  2012  0.526140  0.443622    0.030238         0.300420

In [39]:
#Turnout by age, CPS
#Source: http://www.electproject.org/home/voter-turnout/demographics
# Corrected for response bias
# age_turnout = StringIO("""
#     2016	2012	2008	2004
# 18-29	43.4	40.9	48.4	45.0
# 30-44	56.9	56.2	60.7	59.5
# 45-59	66.2	66.0	69.5	69.0
# 60+	71.4	71.2	71.0	69.7
# """)
# Uncorrected for response bias
age_turnout = StringIO("""
group	2016	2012	2008	2004
17 - 29	46.1	45.0	51.1	49.0
30 - 44	58.8	59.5	61.8	62.4
45 - 59	65.9	66.9	68.5	70.0
60 +	70.3	71.8	70.8	71.4
""")

ageturnout_df = pd.read_csv(age_turnout, sep='\t')
ageturnout_df = pd.melt(ageturnout_df, id_vars=["group"], 
                  var_name="year", value_name="turnout")
ageturnout_df = ageturnout_df.apply(pd.to_numeric, errors='ignore', axis=0)
ageturnout_df['turnout'] = ageturnout_df['turnout']/100

age_df = pd.concat([age2004_df, age2016_df])
age_df = age_df.merge(ageturnout_df, on=['group','year'])

#Add in number in electorate
electorate_df = national_df[['year','num_nation']]
age_df = age_df.merge(electorate_df, on='year', how='left')

age_df['dem_margin'] = age_df['dem_frac'] - age_df['rep_frac']

order = ['demographic','group', 'year','dem_frac','rep_frac','electorate_frac', 'num_nation', 'dem_margin', 'turnout']
age_df = age_df[order]

age_df

demographic    group  year  dem_frac  rep_frac  electorate_frac  \
0          age  17 - 29  2004  0.429017  0.412320         0.123634   
1          age  30 - 44  2004  0.581622  0.373218         0.255757   
2          age  45 - 59  2004  0.420608  0.579392         0.322568   
3          age     60 +  2004  0.445143  0.554857         0.298042   
4          age  17 - 29  2008  0.671394  0.325719         0.141188   
5          age  30 - 44  2008  0.629155  0.350683         0.268001   
6          age  45 - 59  2008  0.562251  0.423448         0.250993   
7          age     60 +  2008  0.504923  0.465264         0.339818   
8          age  17 - 29  2012  0.625693  0.338356         0.139413   
9          age  30 - 44  2012  0.571377  0.368016         0.264247   
10         age  45 - 59  2012  0.512499  0.447605         0.295920   
11         age     60 +  2012  0.526140  0.443622         0.300420   
12         age  17 - 29  2016  0.548326  0.351484         0.173320   
13         age  30 - 44  2016  0.555207  0.345984         0.245604   
14         age  45 - 59  2016  0.446847  0.479965         0.338233   
15         age     60 +  2016  0.475273  0.494703         0.242843   

    num_nation  dem_margin  turnout  
0    122217180    0.016697    0.490  
1    122217180    0.208404    0.624  
2    122217180   -0.158784    0.700  
3    122217180   -0.109714    0.714  
4    131049863    0.345674    0.511  
5    131049863    0.278472    0.618  
6    131049863    0.138803    0.685  
7    131049863    0.039659    0.708  
8    129254383    0.287337    0.450  
9    129254383    0.203361    0.595  
10   129254383    0.064894    0.669  
11   129254383    0.082518    0.718  
12   136670626    0.196842    0.461  
13   136670626    0.209222    0.588  
14   136670626   -0.033118    0.659  
15   136670626   -0.019430    0.703

# Education Data from CPS, ANES 2004-2016

In [24]:
#Turnout by education, CPS
#Source: http://www.electproject.org/home/voter-turnout/demographics
# Corrected for response bias
# Uncorrected for response bias
age_turnout = StringIO("""
group	2016	2012	2008	2004
Less Than High School	34.3	38.0	39.4	39.5
High School Grad	51.6	52.6	54.9	56.4
Some College to College Grad	67.8	68.4	71.5	72.2
Post-Graduate	80.3	81.4	82.7	84.2
""")

edturnout_df = pd.read_csv(age_turnout, sep='\t')
#Just interpolate Some College to Post-Graduate to deal with lack of data
#in ANES survey:
#Either this, or do weighted average later, but it's impossible to do weighted
#average for 2004 to 2012 because you don't know the correct subpopulation.
#Also note that "Less Than High School" and "High School Grad" leaves out those
#that started high school and didn't finish. . . 
#Weight instead based on US overall values:
#https://en.wikipedia.org/wiki/Educational_attainment_in_the_United_States
#Postgrad: 11.77+3.27+1.77 = 16.81%
#Some college: 58.57%
edturnout_df.set_index('group', inplace=True)
#Simple average overestimates:
# edturnout_df.loc['Some College to Post-Graduate'] = (edturnout_df.loc['Some College to College Grad'] + 
#                                                      edturnout_df.loc['Post-Graduate'])/2
#Weighted average:
edturnout_df.loc['Some College to Post-Graduate'] = (edturnout_df.loc['Some College to College Grad']*0.5857 + 
                                                     edturnout_df.loc['Post-Graduate']*0.1681)/(0.5857+0.1681)
edturnout_df.reset_index(inplace=True)
edturnout_df = edturnout_df.loc[[0,1,4]]
edturnout_df = pd.melt(edturnout_df, id_vars=["group"], 
                  var_name="year", value_name="turnout")
edturnout_df = edturnout_df.apply(pd.to_numeric, errors='ignore', axis=0)
edturnout_df['turnout'] = edturnout_df['turnout']/100



edturnout_df

group  year   turnout
0           Less Than High School  2016  0.343000
1                High School Grad  2016  0.516000
2   Some College to Post-Graduate  2016  0.705875
3           Less Than High School  2012  0.380000
4                High School Grad  2012  0.526000
5   Some College to Post-Graduate  2012  0.712990
6           Less Than High School  2008  0.394000
7                High School Grad  2008  0.549000
8   Some College to Post-Graduate  2008  0.739976
9           Less Than High School  2004  0.395000
10               High School Grad  2004  0.564000
11  Some College to Post-Graduate  2004  0.748760

In [25]:
## Calculate 2004-2012 Democratic Margin By Education
## source: ANES time series cumulative:
## http://www.electionstudies.org/studypages/anes_timeseries_cdf/anes_timeseries_cdf.htm

# VCF0110
# DEMOGRAPHICS: Respondent - Education 4-category

# VALID_CODES:
# ------------
# 1.  Grade school or less (0-8 grades)
# 2.  High school (12 grades or fewer, incl. non-college
#       training if applicable)
# 3.  Some college (13 grades or more but no degree; 
#       1948 ONLY: college, no identification of degree
#       status)
# 4.  College or advanced degree (no cases 1948)

#https://nces.ed.gov/fastfacts/display.asp?id=40
#The 6-year graduation rate for first-time, full-time undergraduate students 
#who began seeking a bachelor's degree at a 4-year degree-granting institution 
#in fall 2009 was 59 percent.
# so (finish_college + advanced) = starters*0.59
# some_college = 0.41*starters

ed2004_df = baseanes2004_df[baseanes2004_df['party'].isin([1,2,3])].copy()  #Filter out nonvoters
ed2004_df.rename(columns={'education':'group'},inplace=True)
ed2004_df = ed2004_df.groupby(['year', 'group', 'party']).agg({'weight_postfull':'sum'})
ed2004_df = ed2004_df.unstack(level=-1)
ed2004_df.columns = ed2004_df.columns.droplevel(0)
ed2004_df = ed2004_df.rename_axis(None, axis=1)
ed2004_df.fillna(0, inplace=True)
ed2004_df.reset_index(inplace=True)

#Interpolate to get correct age ranges
labels = {1: "Less Than High School", 2: "High School Grad",
          3: "Some College to Post-Graduate",
          4: "Some College to Post-Graduate"}
ed2004_df['group'].replace(labels, inplace=True)
ed2004_df = ed2004_df.groupby(['year','group']).agg('sum')
ed2004_df.reset_index(inplace=True)

ed2004_df['total'] = ed2004_df[[1,2,3]].sum(axis=1)
ed2004_df.rename(columns={1:'d',2:'r',3:'o'}, inplace=True)

#Divide the number of respondents for each group by total gender respondents
#to get fraction of electorate
sum_df = ed2004_df.groupby('year')['total'].sum()
def electorate_frac(row):
    row['electorate_frac'] = row['total']/sum_df[row['year']]
    return row 
ed2004_df = ed2004_df.apply(electorate_frac, axis=1)
ed2004_df = ed2004_df[ed2004_df['group'].isin(labels.values())]


ed2004_df['dem_frac'] = ed2004_df['d']/ed2004_df['total']
ed2004_df['rep_frac'] = ed2004_df['r']/ed2004_df['total']
ed2004_df['other_frac'] = ed2004_df['o']/ed2004_df['total']

ed2004_df['demographic'] = 'education'
ed2004_df = ed2004_df[['demographic','group','year','dem_frac', 'rep_frac', 
                               'other_frac', 'electorate_frac']]

ed2004_df

demographic                          group  year  dem_frac  rep_frac  \
0    education               High School Grad  2004  0.393690  0.580143   
1    education          Less Than High School  2004  0.652207  0.347793   
2    education  Some College to Post-Graduate  2004  0.479749  0.501125   
4    education               High School Grad  2008  0.582714  0.407605   
5    education          Less Than High School  2008  0.758134  0.241866   
6    education  Some College to Post-Graduate  2008  0.508247  0.466495   
8    education               High School Grad  2012  0.540501  0.439836   
9    education          Less Than High School  2012  0.690355  0.281942   
10   education  Some College to Post-Graduate  2012  0.503933  0.458739   

    other_frac  electorate_frac  
0     0.026167         0.296605  
1     0.000000         0.047057  
2     0.019126         0.656338  
4     0.009682         0.340335  
5     0.000000         0.014612  
6     0.025258         0.641328  
8     0.019663         0.331596  
9     0.027703         0.013457  
10    0.037328         0.644741

In [26]:
## Calculate 2016 Democratic Margin By Education
## source: ANES time series cumulative:
## http://www.electionstudies.org/studypages/anes_timeseries_cdf/anes_timeseries_cdf.htm
# V161270
# What is the highest level of school you have completed or the highest degree you have received?

# FTF Web Total
# 13. Bachelor’s degree (for example: BA, AB, BS)
# 10. Some college but no degree 
# 9. High school graduate- high school diploma or equivalent (for example: GED) 
# 14. Master’s degree (for example: MA, MS, MENG, MED, MSW, MBA) 
# 11. Associate degree in college - occupational/vocational
# program 
# 12. Associate degree in college – academic program
# 8. 12th grade no diploma 
# 16. Doctorate degree (for example: PHD, EDD) 
# 15. Professional school degree (for example: MD, DDS, DVM, LLB, JD)
# 7. 11th grade 
# 6. 10th grade 
# 5. 9th grade 
# 95. Other SPECIFY 
# 4. 7th or 8th grade 
# -9. Refused
anes2016_df = pd.read_csv('./data/anes/anes_timeseries_2016/anes_timeseries_2016_rawdata.txt',
                      sep='|')

columns = ['V160102', 'V161270', 'V162058x']
labels = ['weight_postfull', 'group', 'party']
anes2016_df = anes2016_df[columns]
anes2016_df.rename(columns=dict(zip(columns, labels)), inplace=True)
anes2016_df = anes2016_df[anes2016_df['party'].isin([10,11,12])]  #Filter out nonvoters

ed2016_df = anes2016_df.groupby(['group', 'party']).agg({'weight_postfull':'sum'})
ed2016_df = ed2016_df.unstack(level=-1)
ed2016_df.columns = ed2016_df.columns.droplevel(0)
ed2016_df = ed2016_df.rename_axis(None, axis=1)
ed2016_df.fillna(0, inplace=True)


pairs = [[1,2,3,4],[5,6,7,8,9],[10,11,12,13,14,15,16]]
#Note, I'm including any with some high school in the "High School Grad" category
#because they don't fit anywhere else. . . 
labels = {1: "Less Than High School", 2: "High School Grad",
          3: "Some College to Post-Graduate"}
for i, pair in enumerate(pairs):
    ed2016_df.loc[labels[i+1]] = ed2016_df.loc[pair].sum()

ed2016_df.reset_index(inplace=True)
#Rest are included in sums, so you can remove these:
ed2016_df =  ed2016_df[ed2016_df['group'].isin(labels.values())]
ed2016_df['total'] = ed2016_df[[10,11,12]].sum(axis=1)
ed2016_df.rename(columns={10:'d',11:'r',12:'o'}, inplace=True)
ed2016_df['dem_frac'] = ed2016_df['d']/ed2016_df['total']
ed2016_df['rep_frac'] = ed2016_df['r']/ed2016_df['total']
ed2016_df['other_frac'] = ed2016_df['o']/ed2016_df['total']
ed2016_df.fillna(0, inplace=True)

#Create electorate fraction
ed2016_df['electorate_frac'] = ed2016_df['total'] / ed2016_df['total'].sum()
ed2016_df['year'] = 2016
ed2016_df['demographic'] = 'education'


ed2016_df = ed2016_df[['demographic', 'group','year','dem_frac', 'rep_frac',
                               'other_frac','electorate_frac']]


ed2016_df



/Users/psthomas/miniconda2/envs/datascience3/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2717: DtypeWarning: Columns (790,1129,1131) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


demographic                          group  year  dem_frac  rep_frac  \
19   education          Less Than High School  2016  0.722333  0.277667   
20   education               High School Grad  2016  0.450025  0.486963   
21   education  Some College to Post-Graduate  2016  0.505691  0.422492   

    other_frac  electorate_frac  
19    0.000000         0.008721  
20    0.063012         0.305485  
21    0.071817         0.685794

In [27]:
ed_df = pd.concat([ed2004_df, ed2016_df])

ed_df = ed_df.merge(edturnout_df, on=['year', 'group'])

#Add in number in electorate
electorate_df = national_df[['year','num_nation']]
ed_df = ed_df.merge(electorate_df, on='year', how='left')

ed_df['dem_margin'] = ed_df['dem_frac'] - ed_df['rep_frac']
order = ['demographic','group', 'year','dem_frac','rep_frac','electorate_frac', 'num_nation', 'dem_margin', 'turnout']
ed_df = ed_df[order]

ed_df

demographic                          group  year  dem_frac  rep_frac  \
0    education               High School Grad  2004  0.393690  0.580143   
1    education          Less Than High School  2004  0.652207  0.347793   
2    education  Some College to Post-Graduate  2004  0.479749  0.501125   
3    education               High School Grad  2008  0.582714  0.407605   
4    education          Less Than High School  2008  0.758134  0.241866   
5    education  Some College to Post-Graduate  2008  0.508247  0.466495   
6    education               High School Grad  2012  0.540501  0.439836   
7    education          Less Than High School  2012  0.690355  0.281942   
8    education  Some College to Post-Graduate  2012  0.503933  0.458739   
9    education          Less Than High School  2016  0.722333  0.277667   
10   education               High School Grad  2016  0.450025  0.486963   
11   education  Some College to Post-Graduate  2016  0.505691  0.422492   

    electorate_frac  num_nation  dem_margin   turnout  
0          0.296605   122217180   -0.186452  0.564000  
1          0.047057   122217180    0.304414  0.395000  
2          0.656338   122217180   -0.021377  0.748760  
3          0.340335   131049863    0.175109  0.549000  
4          0.014612   131049863    0.516269  0.394000  
5          0.641328   131049863    0.041751  0.739976  
6          0.331596   129254383    0.100665  0.526000  
7          0.013457   129254383    0.408414  0.380000  
8          0.644741   129254383    0.045194  0.712990  
9          0.008721   136670626    0.444665  0.343000  
10         0.305485   136670626   -0.036938  0.516000  
11         0.685794   136670626    0.083199  0.705875

# Combined Demographic Data

In [40]:
demographic_df = pd.concat([race_df, gender_df, age_df, ed_df])
#demographic_df['year'] = demographic_df['year'].round()
demographic_df = demographic_df.round(decimals=4)

demographic_df

demographic                          group  year  dem_frac  rep_frac  \
0         race             Non-Hispanic White  2012    0.4300    0.5700   
1         race             Non-Hispanic Black  2012    0.9700    0.0300   
2         race                       Hispanic  2012    0.7300    0.2700   
3         race                          Other  2012    0.6100    0.3900   
4         race             Non-Hispanic White  2008    0.4400    0.5600   
5         race             Non-Hispanic Black  2008    1.0000    0.0000   
6         race                       Hispanic  2008    0.7300    0.2700   
7         race                          Other  2008    0.8200    0.1800   
8         race             Non-Hispanic White  2004    0.4200    0.5800   
9         race             Non-Hispanic Black  2004    0.8900    0.1100   
10        race                       Hispanic  2004    0.6100    0.3900   
11        race                          Other  2004    0.3600    0.6400   
12        race             Non-Hispanic White  2016    0.3938    0.5380   
13        race             Non-Hispanic Black  2016    0.8999    0.0638   
14        race                       Hispanic  2016    0.6863    0.2388   
15        race                          Other  2016    0.5732    0.3147   
0       gender                           male  2004    0.4787    0.4919   
1       gender                         female  2004    0.4446    0.5449   
2       gender                           male  2008    0.5050    0.4679   
3       gender                         female  2008    0.5619    0.4237   
4       gender                           male  2012    0.4865    0.4699   
5       gender                         female  2012    0.5494    0.4306   
6       gender                           male  2016    0.4548    0.4690   
7       gender                         female  2016    0.5210    0.4191   
0          age                        17 - 29  2004    0.4290    0.4123   
1          age                        30 - 44  2004    0.5816    0.3732   
2          age                        45 - 59  2004    0.4206    0.5794   
3          age                           60 +  2004    0.4451    0.5549   
4          age                        17 - 29  2008    0.6714    0.3257   
5          age                        30 - 44  2008    0.6292    0.3507   
6          age                        45 - 59  2008    0.5623    0.4234   
7          age                           60 +  2008    0.5049    0.4653   
8          age                        17 - 29  2012    0.6257    0.3384   
9          age                        30 - 44  2012    0.5714    0.3680   
10         age                        45 - 59  2012    0.5125    0.4476   
11         age                           60 +  2012    0.5261    0.4436   
12         age                        17 - 29  2016    0.5483    0.3515   
13         age                        30 - 44  2016    0.5552    0.3460   
14         age                        45 - 59  2016    0.4468    0.4800   
15         age                           60 +  2016    0.4753    0.4947   
0    education               High School Grad  2004    0.3937    0.5801   
1    education          Less Than High School  2004    0.6522    0.3478   
2    education  Some College to Post-Graduate  2004    0.4797    0.5011   
3    education               High School Grad  2008    0.5827    0.4076   
4    education          Less Than High School  2008    0.7581    0.2419   
5    education  Some College to Post-Graduate  2008    0.5082    0.4665   
6    education               High School Grad  2012    0.5405    0.4398   
7    education          Less Than High School  2012    0.6904    0.2819   
8    education  Some College to Post-Graduate  2012    0.5039    0.4587   
9    education          Less Than High School  2016    0.7223    0.2777   
10   education               High School Grad  2016    0.4500    0.4870   
11   education  Some College to Post-Graduate  2016    0.5057    0.4225   

    electorate_frac  num_na

In [41]:
demographic_df.to_csv('./US_Demographic_Presidential_Results_04-16.csv', index=False)